src/projection.jl

```diff
@@ -31,114 +68,269 @@ function Base.show(io::IO, project::ProjectTo{T}) where {T}
     end
 end
 
-# fallback (structs)
-function ProjectTo(x::T) where {T}
-    # Generic fallback for structs, recursively make `ProjectTo`s all their fields
+# Structs
+function generic_projectto(x::T; kw...) where {T}
+    # Generic fallback, recursively make `ProjectTo`s for all their fields
     fields_nt::NamedTuple = backing(x)
-    return ProjectTo{T}(map(ProjectTo, fields_nt))
+    fields_proj = map(fields_nt) do x1
+        if x1 isa Number || x1 isa AbstractArray
+            ProjectTo(x1)
+        else
+            x1
+        end
+    end        
+    # We can't use `T` because if we have `Foo{Matrix{E}}` it should be allowed to make a
+    # `Foo{Diagaonal{E}}` etc. We assume it has a default constructor that has all fields 
+    # but if it doesn't `construct` will give a good error message.
+    wrapT = T.name.wrapper
+    return ProjectTo{wrapT}(; fields_proj..., kw...)
 end
 function (project::ProjectTo{T})(dx::Tangent) where {T}
     sub_projects = backing(project)
     sub_dxs = backing(canonicalize(dx))
-    _call(f, x) = f(x)
-    return construct(T, map(_call, sub_projects, sub_dxs))
-end
-
-# should not work for Tuples and NamedTuples, as not valid tangent types
-function ProjectTo(x::T) where {T<:Union{<:Tuple,NamedTuple}}
-    return throw(
-        ArgumentError("The `x` in `ProjectTo(x)` must be a valid differential, not $x")
-    )
+    maybe_call(f::ProjectTo, x) = f(x)
+    maybe_call(f, x) = f
+    return construct(T, map(maybe_call, sub_projects, sub_dxs))
 end
 
 # Generic
+(::ProjectTo{T})(dx::T) where {T} = dx 
+(::ProjectTo{T})(dx::AbstractZero) where {T} = dx
+(::ProjectTo{T})(dx::NotImplemented) where {T} = dx
+
+ProjectTo() = ProjectTo{Any}()  # trivial case, exists so that maybe_call(f, x) knows what to do
+(x::ProjectTo{Any})(dx) = dx
+
+# Thunks
+(project::ProjectTo)(dx::Thunk) = Thunk(project ∘ dx.f)
+(project::ProjectTo)(dx::InplaceableThunk) = project(dx.val)  # can't update in-place part
 (project::ProjectTo)(dx::AbstractThunk) = project(unthunk(dx))
-(::ProjectTo{T})(dx::T) where {T} = dx  # not always true, but we can special case for when it isn't
-(::ProjectTo{T})(dx::AbstractZero) where {T} = zero(T)
 
-# Number
-ProjectTo(::T) where {T<:Number} = ProjectTo{T}()
+# Zero
+ProjectTo(::AbstractZero) = ProjectTo{AbstractZero}()
+(::ProjectTo{AbstractZero})(dx) = ZeroTangent()
+
+#####
+##### `Base`
+#####
+
+# Bool
+ProjectTo(::Bool) = ProjectTo{AbstractZero}()
```

Comment:
Per above
```suggestion
ProjectTo(::Bool) = ProjectTo{NoTangent}()
```
so we know in the output that the input had no tangent space.
Rather than being disconnected from the problem.
Because those are distinct ideas we do want to preserve